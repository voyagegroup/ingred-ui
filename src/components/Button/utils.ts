import { Theme } from "../../themes";
import { ButtonColor, ButtonColorStyle } from "./Button";

export const getContainerColorStyles = (
  theme: Theme,
): { [P in ButtonColor]: ButtonColorStyle } => ({
  primary: {
    normal: {
      background: theme.palette.primary.main,
      color: theme.palette.text.white,
      boxShadow: theme.shadow["3dShadowPrimary"],
      border: `1px solid ${theme.palette.primary.dark}`,
    },
    hover: {
      background: theme.palette.primary.dark,
      border: "none",
    },
    active: {
      background: theme.palette.primary.dark,
      boxShadow: theme.shadow["3dShadowActive"],
      border: "none",
    },
  },
  basicLight: {
    normal: {
      background: theme.palette.background.default,
      color: theme.palette.black,
      boxShadow: theme.shadow["3dShadowBasic"],
      border: `1px solid ${theme.palette.divider}`,
    },
    hover: {
      background: theme.palette.gray.highlight,
      border: `1px solid ${theme.palette.divider}`,
    },
    active: {
      background: theme.palette.gray.highlight,
      boxShadow: theme.shadow["3dShadowActive"],
      border: `1px solid ${theme.palette.divider}`,
    },
  },
  basicDark: {
    normal: {
      background: theme.palette.basicDark.main,
      color: theme.palette.black,
      boxShadow: theme.shadow["3dShadowBasic"],
      border: `1px solid ${theme.palette.basicDark.deepDark}`,
    },
    hover: {
      background: theme.palette.basicDark.dark,
      border: "none",
    },
    active: {
      background: theme.palette.basicDark.dark,
      boxShadow: theme.shadow["3dShadowBasicActive"],
      border: "none",
    },
  },
  danger: {
    normal: {
      background: theme.palette.danger.main,
      color: theme.palette.text.white,
      boxShadow: theme.shadow["3dShadowDanger"],
      border: `1px solid ${theme.palette.danger.dark}`,
    },
    hover: {
      background: theme.palette.danger.dark,
      border: "none",
    },
    active: {
      background: theme.palette.danger.dark,
      boxShadow: theme.shadow["3dShadowActive"],
      border: "none",
    },
  },
  clear: {
    normal: {
      background: "none",
      color: theme.palette.black,
      boxShadow: "none",
      border: "none",
    },
    hover: {
      background: theme.palette.gray.light,
      border: "none",
    },
    active: {
      background: theme.palette.gray.main,
      boxShadow: "none",
      border: "none",
    },
  },
  primaryPale: {
    normal: {
      background: theme.palette.primaryPale.main,
      color: theme.palette.primaryPale.dark,
      boxShadow: theme.shadow["3dShadowPrimaryPale"],
      border: `1px solid ${theme.palette.primaryPale.light}`,
    },
    hover: {
      background: theme.palette.primaryPale.highlight,
      border: "none",
    },
    active: {
      background: theme.palette.primaryPale.highlight,
      boxShadow: theme.shadow["3dShadowPrimaryPaleActive"],
      border: "none",
    },
  },
});
