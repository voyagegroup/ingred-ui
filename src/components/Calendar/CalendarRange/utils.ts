import dayjs, { Dayjs } from "dayjs";

const isStart = (startDate: Dayjs | null, month: Dayjs, day: number) =>
  startDate?.format("YYYY-MM-DD") ===
  dayjs(new Date(month.year(), month.month(), day)).format("YYYY-MM-DD");

const isEnd = (endDate: Dayjs | null, month: Dayjs, day: number) =>
  endDate?.format("YYYY-MM-DD") ===
  dayjs(new Date(month.year(), month.month(), day)).format("YYYY-MM-DD");

const isBetween = (
  startDate: Dayjs | null,
  endDate: Dayjs | null,
  month: Dayjs,
  day: number,
) =>
  (startDate &&
    endDate &&
    dayjs(new Date(month.year(), month.month(), day)).isAfter(
      startDate.format("YYYY-MM-DD"),
      "day",
    ) &&
    dayjs(new Date(month.year(), month.month(), day)).isBefore(
      endDate.format("YYYY-MM-DD"),
      "day",
    )) ??
  false;

export const getDayState = (
  startDate: Dayjs | null,
  endDate: Dayjs | null,
  month: Dayjs,
  day: number,
) => {
  if (isStart(startDate, month, day)) {
    return "start";
  }
  if (isEnd(endDate, month, day)) {
    return "end";
  }
  if (isBetween(startDate, endDate, month, day)) {
    return "between";
  }
  return "none";
};
