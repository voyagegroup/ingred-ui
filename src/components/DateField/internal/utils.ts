export const getInputWidth = (input: string) =>
  [...input.split("")].reduce((acc, char) => {
    const code = char.charCodeAt(0);
    return acc + (code >= 0x3000 && code <= 0xff60 ? 2 : 1);
  }, 0);
