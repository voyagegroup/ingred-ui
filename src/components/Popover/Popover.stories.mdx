import { Meta, Story, ArgsTable, Canvas } from "@storybook/addon-docs/blocks";
import Popover from "./Popover";

<Meta
  title="Components/Utils/Popover"
  component={Popover}
  parameters={{ docs: { source: { type: "code" } } }}
/>

# Popover

<ArgsTable of={Popover} />

## Samples

<Canvas>
  <Story
    name="Example"
    args={{
      isOpen: false,
      positionPriority: ["top"],
    }}
  >
    {(args) => {
      const [isOpen, setIsOpen] = React.useState(args.isOpen);
      const [buttonElement, setButtonElement] = React.useState(null);
      const handleToggleOpen = () => {
        setIsOpen(!isOpen);
      };
      return (
        <>
          <button ref={setButtonElement} onClick={handleToggleOpen}>
            Click me!
          </button>
          <Popover
            {...args}
            isOpen={isOpen}
            baseElement={buttonElement}
            onClose={handleToggleOpen}
          >
            <div
              style={{
                display: "flex",
                justifyContent: "center",
                alignItems: "center",
                backgroundColor: "white",
                width: "200px",
                height: "200px",
              }}
            >
              hoge
            </div>
          </Popover>
        </>
      );
    }}
  </Story>
</Canvas>
