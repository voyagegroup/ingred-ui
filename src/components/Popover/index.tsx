export { default, PopoverProps } from "./Popover";
