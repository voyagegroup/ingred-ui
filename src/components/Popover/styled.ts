import styled from "styled-components";

import { hexToRgba } from "../../utils/hexToRgba";

export const Container = styled.div`
  background-color: ${({ theme }) => theme.palette.background.default};
  box-shadow: 0px 0px 16px
    ${({ theme }) => hexToRgba(theme.palette.gray.main, 0.4)};
  border-radius: ${({ theme }) => theme.radius}px;
`;
