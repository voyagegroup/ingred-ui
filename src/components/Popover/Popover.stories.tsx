import React from "react";
import { Story } from "@storybook/react/types-6-0";
import { Title, ArgsTable, Stories } from "@storybook/addon-docs";
import Popover, { PopoverProps } from "./Popover";
import Spacer from "../Spacer";
import Flex from "../Flex";
import Button from "../Button";

export default {
  title: "Components/Utils/Popover",
  component: Popover,
  args: {
    isOpen: false,
    positionPriority: ["top"],
  },
  parameters: {
    docs: {
      source: { type: "code" },
      page: () => (
        <>
          <Title />
          <ArgsTable of={Popover} />
          <Stories includePrimary title="Stories" />
        </>
      ),
    },
  },
};

export const Example: Story<PopoverProps> = (args) => {
  const [isOpen, setIsOpen] = React.useState(args.isOpen);
  const [buttonElement, setButtonElement] = React.useState<HTMLElement | null>(
    null,
  );

  const handleToggleOpen = () => {
    setIsOpen(!isOpen);
  };
  return (
    <>
      <Spacer mt={16} />
      <Flex display="flex" justifyContent={"center"}>
        <Button ref={setButtonElement} inline={true} onClick={handleToggleOpen}>
          Click me!
        </Button>
      </Flex>
      <Popover
        {...args}
        isOpen={isOpen}
        baseElement={buttonElement}
        onClose={handleToggleOpen}
      >
        <div
          style={{
            display: "flex",
            justifyContent: "center",
            alignItems: "center",
            backgroundColor: "white",
            width: "100px",
            height: "100px",
          }}
        >
          hoge
        </div>
      </Popover>
    </>
  );
};
