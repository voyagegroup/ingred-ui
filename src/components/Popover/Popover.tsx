import { Placement } from "@floating-ui/core";
import * as React from "react";
import { useMergeRefs } from "../../hooks/useMergeRefs";
import * as Styled from "./styled";
import {
  useFloating,
  autoUpdate,
  offset as floatingOffset,
  shift,
  FloatingFocusManager,
  useRole,
  useInteractions,
  autoPlacement,
  flip,
} from "@floating-ui/react";
import Modal, { ModalProps } from "../Modal";
import { usePlacement, AutoPlacement } from "../../hooks/usePlacement";

export type PopoverProps = React.ComponentPropsWithoutRef<"div"> & {
  /**
   * If `false`, children becomes `visibility: hidden`.
   */
  isOpen?: boolean;
  onClose?: ModalProps["onClose"];
  baseElement: HTMLElement | null;
  /**
   * Define priority of position. Please check [this](https://floating-ui.com/docs/tutorial#placements).
   * If not specified, it will be auto.
   */
  positionPriority?: (Placement | AutoPlacement)[];
  offset?: [number, number];
  /**
   * props of [Modal](/?path=/docs/components-utils-modal)
   */
  modalProps?: Partial<ModalProps>;
  children: React.ComponentElement<HTMLElement, any>;
};

const Popover = React.forwardRef<HTMLDivElement, PopoverProps>(function Popover(
  {
    isOpen = true,
    onClose,
    baseElement,
    offset = [0, 0],
    positionPriority = ["auto"],
    modalProps,
    children,
    ...rest
  },
  ref,
) {
  const { placements, isAuto } = usePlacement(positionPriority);
  const {
    x,
    y,
    refs: floatingRef,
    strategy,
    context,
  } = useFloating({
    elements: {
      reference: baseElement,
    },
    placement: placements[0],
    open: isOpen,
    middleware: [
      positionPriority.length > 0 && !isAuto
        ? flip({
            padding: 24,
            fallbackPlacements: placements,
          })
        : autoPlacement(),
      floatingOffset({
        mainAxis: offset[1],
        crossAxis: offset[0],
      }),
      shift({
        mainAxis: false,
      }),
    ],
    whileElementsMounted: autoUpdate,
  });

  const role = useRole(context);
  const { getFloatingProps } = useInteractions([role]);

  const refs = useMergeRefs<HTMLDivElement>(ref, floatingRef.setFloating);

  return (
    <Modal
      isOpen={isOpen}
      backdropProps={{ invisible: true }}
      {...modalProps}
      onClose={onClose}
    >
      <FloatingFocusManager context={context} modal={false} initialFocus={-1}>
        <Styled.Container
          ref={refs}
          style={{
            position: strategy,
            top: y,
            left: x,
          }}
          {...getFloatingProps()}
          {...rest}
        >
          {children}
        </Styled.Container>
      </FloatingFocusManager>
    </Modal>
  );
});

export default Popover;
