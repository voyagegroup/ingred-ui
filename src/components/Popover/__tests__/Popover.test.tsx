import * as React from "react";
import "@testing-library/jest-dom/extend-expect";
import { cleanup } from "@testing-library/react";
import Popover from "..";
import { renderWithThemeProvider } from "../../../utils/renderWithThemeProvider";

jest.mock("react-dom", () => {
  const original = jest.requireActual("react-dom");
  return {
    ...original,
    createPortal: (node: any) => node,
  };
});

describe("Popover component testing", () => {
  afterEach(cleanup);

  test("Popover", () => {
    const { asFragment } = renderWithThemeProvider(
      <Popover baseElement={null}>
        <div>hoge</div>
      </Popover>,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
