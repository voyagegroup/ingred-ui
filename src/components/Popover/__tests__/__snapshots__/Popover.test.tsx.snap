// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popover component testing Popover 1`] = `
<DocumentFragment>
  <div
    class="sc-gsDKAQ gzCcYc"
  >
    <div
      class="sc-hKwDye dOfRKQ fade-transition-appear fade-transition-appear-active"
    >
      <div
        class="sc-dkPtRN jXTMGN"
      />
    </div>
    <div
      class="sc-bdvvtL wkcbx"
      style="position: absolute; left: 0px; top: 0px;"
    >
      <div>
        hoge
      </div>
    </div>
  </div>
</DocumentFragment>
`;
